CIP-0084? | Ledger evolution (meta) (#456)
cardano-foundation/CIPs: CIP-????/README.md

**Hard fork** - A *hard fork* is a change to the protocol (not restricted to the ledger)
resulting in a single new block definition becoming valid.

Alternatively, a hard fork is a backwards incompatible change for both
block producers and block validators.

**Soft fork** - A *soft fork* is a change to the protocol (not restricted to the ledger)
resulting in fewer blocks being valid.

Alternatively, a soft fork is a backwards incompatible change for
block producers, but a backwards compatible change for block validators.

### Serialization

Transactions and blocks are serialized with
[CBOR](https://www.rfc-editor.org/rfc/rfc7049)
and specified with
[CDDL file](https://www.rfc-editor.org/rfc/rfc8610).

Serialization changes to the ledger are discussed in
[CIP-80](https://github.com/cardano-foundation/CIPs/pull/372).

Note that the serialisation format of the ledger state is unspecified and left as an
implementation detail (unlike the format of blocks).

### The ledger-script interface

The ledger and Plutus scripts have a common interface, described in [CIP-35].
CIPs relating to this inteface are relevant to both the ledger and to the Plutus CIP categories.

> 

Reviewer: This section could also use a brief reference like what @michaelpj put in CIP-0035 about categorising CIPs which authors might be tempted to categorise as both `plutus` and `ledger`: https://github.com/cardano-foundation/CIPs/tree/master/CIP-0035#scope

The pointer to CIP-00035 will work for many users, but some authors & third parties will be helped by quoting or summarising this paragraph from it:
> Additionally, there is significant overlap with the Ledger category around the ledger-script interface and the protocol parameters. CIPs which change these parts of Cardano should generally use the Plutus category and not the Ledger category, although the Editors may ask the Ledger reviewers to comment.

Author: great idea, thanks!

  a formal ledger specification.
* Some changes may require getting an expert opinion.
  For example, changes to the incentives model (such as changing the reward calculation)
  should include a proper game theoretic analysis.

## Rationale

### There is only one implementation, why limit the scope of ledger CIPs in this way?

Even though there is currently only one implementation, this provides us with a clear
definition of what is essential to the ledger.
It also provides a clear path for future implementations.

### Why is the specification vague about the role of ledger events in the CIP process?

This decision should be left to the community as more use cases emerge.

### Why is familiarity with the formal specifications required?

It is not always clear which seemingly small details can make a large difference
to the many consumers of the ledger.
It is better that the CIP process achieve consensus on all the details than for
these decisions to be made during the implementation phase.

[CIP-35]: https://github.com/cardano-foundation/CIPs/tree/master/CIP-0035
[CIP-59]: https://github.com/cardano-foundation/CIPs/tree/master/CIP-0059

> [CIP-1694]: https://github.com/cardano-foundation/CIPs/tree/master/CIP-1694

Reviewer: @JaredCorduan maybe you were planning to add the required _Path to Active_ here & this is still a draft?  If not prepared yet, this could again resemble CIP-0035 with "Ledger" substituted for "Plutus" ... i.e. it's accepted when the Ledger team accepts the CIP, and implemented when the Ledger team implements it ... or something more articulated if that would suit you better 😎

https://github.com/cardano-foundation/CIPs/tree/master/CIP-0035#path-to-active

Author: great, I'm happy to keep consistent wording with CIP-35.

> [CIP-1694]: https://github.com/cardano-foundation/CIPs/tree/master/CIP-1694

Reviewer: ... and another obligatory section: your chose software licensing model and/or link, as per https://github.com/cardano-foundation/CIPs/tree/master/CIP-0001#licensing

## Specification

### What merits a ledger CIP?

The criterion for deciding if a change to the ledger merits a CIP is as follows:
changes to the ledger require going through the CIP process whenever
every implementation of the Cardano ledger needs to be standardized on the details.

Bug fixes are an exception to this criterion, they do not merit a CIP except in the case
that the fix is substantially complicated.
A "bug fix" is a change to behavior where:
- The implemented behavior does not match the specification; or
- The specified behavior is clearly wrong (in the judgment of relevant experts)

Serialization changes are another possible exception to the criterion.
Many serialization changes can be handled as a part of the normal development process
without the need for a CIP.
Dramatic changes to the serialization, however, may benefit from the CIP process.

The ledger rules MUST be standardized in order for consensus to be maintained,
but things like the ledger events are more open to debate.

Changes to the protocol parameter values do not require a CIP since they are
a governance issue (see [CIP-1694]).

> ### Ledger Core CIP registry

Reviewer 2: I think this isn't needed any more since there's a category that does the job of aggregating things.

|   32 | Inline datums                               | Active   |
|   33 | Reference scripts                           | Active   |
|   59 | Terminology Surrounding Core Features       | Active   |
|   80 | Transaction Serialization Deprecation Cycle | Proposed |
| 1694 | A proposal for entering the Voltaire phase  | Proposed |

### Expectations for ledger CIPs

* Familiarity with the existing ledger specifications is required to propose changes to the ledger.
* The CIP specifications for ledger CIPs must be sufficiently detailed for inclusion in
  a formal ledger specification.
* Some changes may require getting an expert opinion.
  For example, changes to the incentives model (such as changing the reward calculation)
  should include a proper game theoretic analysis.

## Rationale

### There is only one implementation, why limit the scope of ledger CIPs in this way?

Even though there is currently only one implementation, this provides us with a clear
definition of what is essential to the ledger.
It also provides a clear path for future implementations.

### Why is the specification vague about the role of ledger events in the CIP process?

> This decision should be left to the community as more use cases emerge.

Reviewer 2: I still think ledger events should pretty clearly be "out". They don't affect the operation of the Cardano network. A competing implementation could omit them and be perfectly correct.

resulting in fewer blocks being valid.

Alternatively, a soft fork is a backwards incompatible change for
block producers, but a backwards compatible change for block validators.

### Serialization

Transactions and blocks are serialized with
[CBOR](https://www.rfc-editor.org/rfc/rfc7049)
and specified with
[CDDL file](https://www.rfc-editor.org/rfc/rfc8610).

Serialization changes to the ledger are discussed in
[CIP-80](https://github.com/cardano-foundation/CIPs/pull/372).

Note that the serialisation format of the ledger state is unspecified and left as an
implementation detail (unlike the format of blocks).

### The ledger-script interface

The ledger and Plutus scripts have a common interface, described in [CIP-35].
CIPs relating to this inteface are relevant to both the ledger and to the Plutus CIP categories.

Additionally, there is significant overlap with the Ledger category around the ledger-script
interface and the protocol parameters.
CIPs which change these parts of Cardano should generally use the Plutus category and not the
Ledger category, although the Editors may ask the Ledger reviewers to comment.

> ## Specification

Author: I see that CIP-35 has a [Reviewers](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0035#the-plutus-reviewers) section now. I hadn't planned on adding a table like that, I suspect that the folks I would put in such a table would end up review ledger CIPs anyway. Should I re-consider? cc @michaelpj

Reviewer 2: I'm not sure. I put it there because the new CIP-001 somewhat implicitly has the concept of project maintainers/reviewers who are the people who get called upon to review CIPs in that category, and I thought it was better to make it explicit who that was.

---

> CIP: ?

Reviewer 3: ```suggestion
CIP: 84
```

Title: Cardano Ledger Evolution

> Status: Proposed

Reviewer 3: ```suggestion
Status: Active
```

Category: Meta

> Authors: Jared Corduan <jared.corduan@iohk.io>

Reviewer 3: ```suggestion
Authors: 
  - Jared Corduan <jared.corduan@iohk.io>
```

Implementors: N/A

> Created: 2023-01-30

Reviewer 3: ```suggestion
Created: 2023-01-30
Discussions: 
  - https://github.com/cardano-foundation/CIPs/pull/456
```

License: CC-BY-4.0
---

> # Evolution of the Cardano ledger

Reviewer 3: ```suggestion
```

## Abstract

This CIP provides guidance for future CIPs concerning the Cardano ledger.

> ## Motivation

Reviewer 3: ```suggestion
## Motivation: why is this CIP necessary?
```

The ledger is responsible for processing transactions and updating the shared state of the network.
It also processes block headers and handles the state transformation from one epoch to the next
(e.g. computing the staking rewards).

Most of the state maintained by the ledger relates to the
[Extended UTxO accounting model](https://iohk.io/en/research/library/papers/the-extended-utxo-model/) and
support for [Ouroboros](https://iohk.io/en/research/library/papers/ouroboros-a-provably-secure-proof-of-stake-blockchain-protocol/),
the proof-of-stake consensus mechanism used in Cardano.

This CIP aims to give guidance for future CIPs related to the ledger,
making it a registered category of the CIP process[^1].
[^1]: See [CIP-1](https://github.com/cardano-foundation/CIPs/blob/cip-cps-rework/CIP-0001/README.md#categories).
While nothing new is added to the usual CIP process,
expectations for ledger CIPs are made explicit and some background information is provided.

Many thanks to Arnaud Bailly and Michael Peyton Jones for all their help reviewing and providing
feedback on the first versions of this CIP.

> ## Background

Reviewer 3: ```suggestion
### Specification
```

resulting in fewer blocks being valid.

Alternatively, a soft fork is a backwards incompatible change for
block producers, but a backwards compatible change for block validators.

### Serialization

Transactions and blocks are serialized with
[CBOR](https://www.rfc-editor.org/rfc/rfc7049)
and specified with
[CDDL file](https://www.rfc-editor.org/rfc/rfc8610).

Serialization changes to the ledger are discussed in
[CIP-80](https://github.com/cardano-foundation/CIPs/pull/372).

Note that the serialisation format of the ledger state is unspecified and left as an
implementation detail (unlike the format of blocks).

### The ledger-script interface

The ledger and Plutus scripts have a common interface, described in [CIP-35].
CIPs relating to this inteface are relevant to both the ledger and to the Plutus CIP categories.

Additionally, there is significant overlap with the Ledger category around the ledger-script
interface and the protocol parameters.
CIPs which change these parts of Cardano should generally use the Plutus category and not the
Ledger category, although the Editors may ask the Ledger reviewers to comment.

> ## Specification

Reviewer 3: ```suggestion
```

### What merits a ledger CIP?

The criterion for deciding if a change to the ledger merits a CIP is as follows:
changes to the ledger require going through the CIP process whenever
every implementation of the Cardano ledger needs to be standardized on the details.

Bug fixes are an exception to this criterion, they do not merit a CIP except in the case
that the fix is substantially complicated.
A "bug fix" is a change to behavior where:
- The implemented behavior does not match the specification; or
- The specified behavior is clearly wrong (in the judgment of relevant experts)

Serialization changes are another possible exception to the criterion.
Many serialization changes can be handled as a part of the normal development process
without the need for a CIP.
Dramatic changes to the serialization, however, may benefit from the CIP process.

The ledger rules MUST be standardized in order for consensus to be maintained,
but things like the ledger events are more open to debate.

Changes to the protocol parameter values do not require a CIP since they are

> a governance issue (see [CIP-1694]).

Reviewer 3: I think this section would benefit from a little bit of rework to state more clearly what is a potential candidate for a `Ledger` CIP, and what isn't. I'd suggest to make two lists with bullet points. 

I would be keen on seeing a mention of the RSS here (in either of the list).

cc @JaredCorduan

Author: @KtorZ do you mean that the content is fine, but that it would help to use bullets to make it more visually clear?

What RSS are you referring to?

Reviewer 3: > @JaredCorduan do you mean that the content is fine, but that it would help to use bullets to make it more visually clear?

Yes, the suggestion is purely _structural_. From the perspective of a contributor, I'd like to easily grasp in one look what are suitable topics for a ledger CIP and what aren't.

> What RSS are you referring to?

I am referring to anything that touches the rewards and incentives schemes. This is a common topic of discussion in CIPs; ranging from suggestions of new values for protocol parameters to changes in the reward equations (in particular w.r.t to the pledge influence) or complete redesign of the incentives. 

Those CIPs have historically gotten very little attention and are hard to move forward. I'd appreciate clarity as to whether the `Ledger` category now cover them, or not. If the intent is to cover this area, then I'd also love the process to be specified in more detail in this proposal. So far, it seems that RSS questions have required the involvement of IO research but the process hasn't always / ever been smooth as there's little availability from the research side. 

Plus, all CIPs in this situation never got a clear answer from IO / the ledger team regarding their "acceptance". That's why from an edition perspective, we've recently decided to simply mark all of them as _Inactive_ because we don't have any procedure by which we can propose them and move them forward. 

If this proposal still doesn't address the case of RSS proposals, then let's state it clearly here too.

Author: > I'd like to easily grasp in one look what are suitable topics for a ledger CIP and what aren't.

I definitely appreciate your help to make this more readable and friendly. Really there is just one rule:

```
every implementation of the Cardano ledger needs to be standardized on the details.
```

The paragraphs below this, that I think you are suggesting making into a bullet list, are just elaborations on this and exceptions. Maybe I could bold the main criterion sentence? Would adding bullets to the other things make the relevance of the main criterion seem less apparent?

Author: regarding the rewards and incentives schemes, my hope was that it would be clear that indeed they are in scope:

```
every implementation of the Cardano ledger needs to be standardized on the details.
```

The caveat is listed below:

```
 For example, changes to the incentives model (such as changing the reward calculation)
  should include a proper game theoretic analysis.
```

Of course, as you can see, I've given no guidance on what is "proper" or who decides. 

> Those CIPs have historically gotten very little attention and are hard to move forward

I think it is more that there are only so many things we can do at once, and we have been focused on babbage and voltaire.

Reviewer: @JaredCorduan yes I think it does appear that the need to categorise RSS proposals _somehow_ would be satisfied by giving them a `Ledger` category... while the comment you quote directly above also indicates that the consideration of these proposals will generally depend on expertise & review outside the Ledger team (i.e. a "disclaimer" so that their responsibility would not fall exclusively upon them).

So in my opinion this question has no reason to block your proposal as it currently stands... but I'll leave this conversation open so we can see it in our CIP meeting (1 hour from now). (cc @KtorZ)

### Expectations for ledger CIPs

* Familiarity with the existing ledger specifications is required to propose changes to the ledger.
* The CIP specifications for ledger CIPs must be sufficiently detailed for inclusion in
  a formal ledger specification.
* Some changes may require getting an expert opinion.
  For example, changes to the incentives model (such as changing the reward calculation)
  should include a proper game theoretic analysis.

### The Ledger reviewers

The following table gives the current set of reviewers for Ledger CIPs.

| Name          | Email                 | GitHub username |
|---------------|-----------------------|-----------------|

> | Jared Corduan | jared.corduan@iohk.io | JaredCorduan    |

Reviewer 3: ```suggestion
| Jared Corduan | jared.corduan@iohk.io | @JaredCorduan  |
```

> | Andre Knispel | andre.knispel@iohk.io | WhatisRT        |

Reviewer 3: ```suggestion
| Andre Knispel | andre.knispel@iohk.io | @WhatisRT        |
```

> 

Reviewer 3: I think the "Specification" would benefit from a short section about what it takes for a ledger CIP to _become active_. Are you expecting people to submit PRs to the ledger, or would you rather avoid people doing so (and have only the core maintainers do the work that's agreed on CIPs)?

Author: > Are you expecting people to submit PRs to the ledger, or would you rather avoid people doing so (and have only the core maintainers do the work that's agreed on CIPs)?

Something in the middle, I think. This involves perhaps a big topic that we should discuss.

Features do not get implemented in isolation in the ledger. We collect up a bunch of features, see how they interact, design a specification for a new ledger era, and then implement the era. I actually wrote up an earlier draft of this CIP that was much different than what I've submitted, where "ledger era" was its own ledger sub-type. In the end, I talked myself out of this since it could be argued that this is more of the release process. The ledger team would of course love help with PRs, but it would need to match our existing workflow, which is to build out the code once the era is settled. I'm not sure how the bundling should be involved in the CIP process.

Author: > short section about what it takes for a ledger CIP to become active

I could probably use some guidance from the CIP committee as to how to organize this. These are the stages in my mind:

* everyone agrees it is a good change
* it is selected for inclusion in some ledger era
* it is added to the formal spec
* it is implemented
* the hard fork happens and it is live

> ## Rationale

Reviewer 3: ```suggestion
## Rationale: how does this CIP achieve its goals?
```

every implementation of the Cardano ledger needs to be standardized on the details.

Bug fixes are an exception to this criterion, they do not merit a CIP except in the case
that the fix is substantially complicated.
A "bug fix" is a change to behavior where:
- The implemented behavior does not match the specification; or
- The specified behavior is clearly wrong (in the judgment of relevant experts)

Serialization changes are another possible exception to the criterion.
Many serialization changes can be handled as a part of the normal development process
without the need for a CIP.
Dramatic changes to the serialization, however, may benefit from the CIP process.

The ledger rules MUST be standardized in order for consensus to be maintained,
but things like the ledger events are more open to debate.

Changes to the protocol parameter values do not require a CIP since they are
a governance issue (see [CIP-1694]).

### Expectations for ledger CIPs

* Familiarity with the existing ledger specifications is required to propose changes to the ledger.
* The CIP specifications for ledger CIPs must be sufficiently detailed for inclusion in
  a formal ledger specification.

> * Some changes may require getting an expert opinion.
>   For example, changes to the incentives model (such as changing the reward calculation)
>   should include a proper game theoretic analysis.

Reviewer: ```suggestion
* Though proposals can be accepted solely on the basis of peer and Ledger team review, some areas (e.g. changes to the incentives model) might only considered ready for implementation if accompanied by an opinion from an expert designated by the implementor (e.g. with a proper game theoretic analysis).
```
Note the parallel example (two e.g.'s here, like logical propositions, assumed to relate to each other) was argued for by @JaredCorduan somewhere in the CIP-1694 discussion but I can't find the reference to that comment. 😅

Author: I :heart: the suggestion!
